spec/features/masquerade_feature_spec.rb

```diff
@@ -0,0 +1,101 @@
+    expect(page).to have_css('.sign-in-container')
+    expect(page).not_to have_content('Sign Out')
+  end
+
+  def expect_masquerading_to_fail(educator)
+    sign_in_as(educator)
+    simulate_clicking_become_link(page, pals.uri.id)
+    expect_not_to_be_masquerading(page)
+  end
+
+  # This is all a manual workaround for not being able to click on links that use
+  # jquery_ujs in RackTest (there's no JS running to interpret the click and translate it
+  # into a post request).  If we just used `visit` directly, it'd make a get request and
+  # not work.
+  #
+  # This doesn't actually test the content of the link on the page.
+  def simulate_clicking_become_link(page, target_educator_email)
+    page.driver.post admin_masquerade_become_url(masquerading_educator_id: pals.shs_jodi.id)
+
+    # I'm not sure why, but everything to this point has been http, and this
+    # redirects to https.  Also not sure why Capybara doesn't follow this (perhaps
+    # because we're manually make posting requests to RackTest.  Either way,
+    # this manually follows the redirect and makes another post to the same path at https.
+    page.driver.post page.driver.response.location
+
+    # And one more time we manually follow the expected redirect (in this case it's back to
+    # the home page).
+    visit page.driver.response.location
+  end
+
+  it 'works for Uri to become any user' do
+    sign_in_as(pals.uri)
+    simulate_clicking_become_link(page, pals.shs_jodi.id)
+    expect_to_be_masquerading_as(page, pals.shs_jodi)
+  end
+
+  it 'does not work for Jodi to become any user' do
+    sign_in_as(pals.shs_jodi)
+    simulate_clicking_become_link(page, pals.uri.id)
+    expect_not_to_be_masquerading(page)
+    sign_out
+    expect_to_be_logged_out(page)
+  end
+
+  describe 'blocks all other TestPals' do
+    it 'rich_districtwide' do expect_masquerading_to_fail(pals.rich_districtwide) end
+    it 'healey_vivian_teacher' do expect_masquerading_to_fail(pals.healey_vivian_teacher) end
+    it 'healey_ell_teacher' do expect_masquerading_to_fail(pals.healey_ell_teacher) end
+    it 'healey_sped_teacher' do expect_masquerading_to_fail(pals.healey_sped_teacher) end
+    it 'healey_laura_principal' do expect_masquerading_to_fail(pals.healey_laura_principal) end
+    it 'healey_sarah_teacher' do expect_masquerading_to_fail(pals.healey_sarah_teacher) end
+    it 'west_marcus_teacher' do expect_masquerading_to_fail(pals.west_marcus_teacher) end
+    it 'shs_jodi' do expect_masquerading_to_fail(pals.shs_jodi) end
+    it 'shs_bill_nye' do expect_masquerading_to_fail(pals.shs_bill_nye) end
+    it 'shs_ninth_grade_counselor' do expect_masquerading_to_fail(pals.shs_ninth_grade_counselor) end
+    it 'shs_hugo_art_teacher' do expect_masquerading_to_fail(pals.shs_hugo_art_teacher) end
+    it 'shs_fatima_science_teacher' do expect_masquerading_to_fail(pals.shs_fatima_science_teacher) end
+  end
+
+  # it 'does not work for any other users to masquerade as anyone else' do
```

Comment:
This test should be un-commented.

Reply:
I'm not sure if we should do this so left it here so we could discuss, but the double revert lost the comment :)

https://github.com/studentinsights/studentinsights/pull/1749#discussion_r192570372
> This seems most thorough, but takes ~30 seconds to run through. A compromise is the lines above, which explicitly tests those users trying to masquerade as the highest-permission user, which is a bit faster and gives feedback on each one.

What do you think is the right tradeoff here?

Reply:
Ah, gotcha! Yeah, I see why this would be really slow with the nested looping across `Educator.all`. 

Two thoughts. First, I think it would be good to bring back the outer Educator loop instead of naming and testing each educator individually. That's because we sometimes add TestPal educators. Looping means that each new TestPal will be covered automatically and there's no need to remember to update the spec.

In terms of the inner loop, I agree that testing for masquerade-as-Uri is a good idea. One compromise would be to add two tests to the inner loop, one against masquerade-as-Uri and one against masquerade-as-another-user. That way if something changes about the Uri TestPal that might screw with this test (unlikely but just trying to think about future-proofing), at least it's not the only test role we're testing against.

Reply:
@alexsoble sure, sounds good!